src/main/java/net/snowflake/client/core/SFArrowResultSet.java

```diff
@@ -763,6 +788,40 @@ public BigDecimal getBigDecimal(int columnIndex, int scale) throws SFException {
     return bigDec == null ? null : bigDec.setScale(scale, RoundingMode.HALF_UP);
   }
 
+  public ArrowBatches getArrowBatches() {
+    if (readingMode == ReadingMode.ROW_MODE) {
+      logger.warn("Cannot read arrow batches after next() was called.");
+      return null;
+    }
+    readingMode = ReadingMode.BATCH_MODE;
+    return new SFArrowBatchesIterator();
+  }
+
+  private class SFArrowBatchesIterator implements ArrowBatches {
+    private boolean firstFetched = false;
+
+    @Override
+    public long getRowCount() throws SQLException {
+      return resultSetSerializable.getRowCount();
+    }
+
+    @Override
+    public boolean hasNext() {
+      return nextChunkIndex < chunkCount || !firstFetched;
+    }
+
+    @Override
+    public ArrowBatch next() throws SQLException {
+      if (!firstFetched) {
+        firstFetched = true;
+        return currentChunkIterator.getChunk().getArrowBatch(SFArrowResultSet.this);
+      } else {
+        nextChunkIndex++;
+        return fetchNextChunk().getArrowBatch(SFArrowResultSet.this);
```

Comment:
I think the issue here is a failure to load an arrow chunk. It's quite different to an iterator, in the fact that this may happen even though hasNext() returned true. To preserve the contract `hasNext() = true` => `next()` does not throw,  we would need to try the download in `hasNext()`. Then either `hasNext()` throws, or it returns `false` just as if the iteration had finished correctly, neither of which is perfect. In summary, I don't think there is a perfect solution here.

Reply:
I get it, let's leave it as is then